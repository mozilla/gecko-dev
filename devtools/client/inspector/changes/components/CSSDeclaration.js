/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { PureComponent } = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");

class CSSDeclaration extends PureComponent {
  static get propTypes() {
    return {
      property: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
      className: PropTypes.string,
    };
  }

  static get defaultProps() {
    return {
      className: "",
    };
  }

  render() {
    const { property, value, className } = this.props;

    return dom.div({ className: `declaration ${className}` },
      dom.span({ className: "declaration-name theme-fg-color5"}, property),
      ":",
      dom.span({ className: "declaration-value theme-fg-color1"}, value),
      ";"
    );
  }
}

module.exports = CSSDeclaration;
