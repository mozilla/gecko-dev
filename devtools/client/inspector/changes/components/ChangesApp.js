/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { createFactory, PureComponent } = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const { connect } = require("devtools/client/shared/vendor/react-redux");

const CSSDeclaration = createFactory(require("./CSSDeclaration"));
const { getStr } = require("../utils/l10n");

class ChangesApp extends PureComponent {
  static get propTypes() {
    return {
      // Redux state slice assigned to Track Changes feature; passed as prop by connect()
      changes: PropTypes.object.isRequired,
    };
  }

  constructor(props) {
    super(props);
    // In the Redux store, all rules exist in a collection at the same level of nesting.
    // Parent rules come before child rules. Parent/child dependencies are set
    // via parameters in each rule pointing to the corresponding rule ids.
    //
    // To render rules, we traverse the descendant rule tree and render each child rule
    // found. This means we get into situations where we can render the same rule multiple
    // times: once as a child of its parent and once standalone.
    //
    // By keeping a log of rules previously rendered we prevent needless multi-rendering.
    this.renderedRules = [];
  }

  renderDeclarations(remove = [], add = []) {
    const removals = remove
      // Sorting changed declarations in the order they appear in the Rules view.
      .sort((a, b) => a.index > b.index)
      .map(({property, value, index}) => {
        return CSSDeclaration({
          key: "remove-" + property + index,
          className: "level diff-remove",
          property,
          value,
        });
      });

    const additions = add
      // Sorting changed declarations in the order they appear in the Rules view.
      .sort((a, b) => a.index > b.index)
      .map(({property, value, index}) => {
        return CSSDeclaration({
          key: "add-" + property + index,
          className: "level diff-add",
          property,
          value,
        });
      });

    return [removals, additions];
  }

  renderRule(ruleId, rule, rules) {
    const selector = rule.selector;

    if (this.renderedRules.includes(ruleId)) {
      return null;
    }

    // Mark this rule as rendered so we don't render it again.
    this.renderedRules.push(ruleId);

    let diffClass = "";
    if (rule.changeType === "rule-add") {
      diffClass = "diff-add";
    } else if (rule.changeType === "rule-remove") {
      diffClass = "diff-remove";
    }

    return dom.div(
      {
        key: ruleId,
        className: "rule",
      },
      dom.div(
        {
          className: `level selector ${diffClass}`,
          title: selector,
        },
        selector,
        dom.span({ className: "bracket-open" }, "{")
      ),
      // Render any nested child rules if they are present.
      rule.children.length > 0 && rule.children.map(childRuleId => {
        return this.renderRule(childRuleId, rules[childRuleId], rules);
      }),
      // Render any changed CSS declarations.
      this.renderDeclarations(rule.remove, rule.add),
      dom.div({ className: `level bracket-close ${diffClass}` }, "}")
    );
  }

  renderDiff(changes = {}) {
    // Render groups of style sources: stylesheets and element style attributes.
    return Object.entries(changes).map(([sourceId, source]) => {
      const href = source.href || `inline stylesheet #${source.index}`;
      const rules = source.rules;

      return dom.details(
        {
          key: sourceId,
          className: "source devtools-monospace",
          open: true,
        },
        dom.summary(
          {
            className: "href",
            title: href,
          },
          href),
        // Render changed rules within this source.
        Object.entries(rules).map(([ruleId, rule]) => {
          return this.renderRule(ruleId, rule, rules);
        })
      );
    });
  }

  renderEmptyState() {
    return dom.div({ className: "devtools-sidepanel-no-result" },
      dom.p({}, getStr("changes.noChanges")),
      dom.p({}, getStr("changes.noChangesDescription"))
    );
  }

  render() {
    // Reset log of rendered rules.
    this.renderedRules = [];
    const hasChanges = Object.keys(this.props.changes).length > 0;

    return dom.div(
      {
        className: "theme-sidebar inspector-tabpanel",
        id: "sidebar-panel-changes",
      },
      !hasChanges && this.renderEmptyState(),
      hasChanges && this.renderDiff(this.props.changes)
    );
  }
}

module.exports = connect(state => state)(ChangesApp);
