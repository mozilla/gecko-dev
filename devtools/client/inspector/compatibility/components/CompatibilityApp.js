/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { connect } = require("devtools/client/shared/vendor/react-redux");
const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");

const Types = require("../types");

const IssueList = createFactory(require("./IssueList"));

class CompatibilityApp extends PureComponent {
  static get propTypes() {
    return {
      selectedNodeIssues: PropTypes.arrayOf(PropTypes.shape(Types.issue))
        .isRequired,
    };
  }

  render() {
    const { selectedNodeIssues } = this.props;

    return dom.div(
      {
        className: "theme-sidebar inspector-tabpanel",
      },
      selectedNodeIssues.length
        ? IssueList({ issues: selectedNodeIssues })
        : null
    );
  }
}

const mapStateToProps = state => {
  return {
    selectedNodeIssues: state.compatibility.selectedNodeIssues,
  };
};
module.exports = connect(mapStateToProps)(CompatibilityApp);
