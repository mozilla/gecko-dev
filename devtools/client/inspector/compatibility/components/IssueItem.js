/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const Services = require("Services");
const { PureComponent } = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");

const Types = require("../types");

class IssueItem extends PureComponent {
  static get propTypes() {
    return {
      ...Types.issue,
    };
  }

  _isTesting() {
    return Services.prefs.getBoolPref("devtools.testing", false);
  }

  render() {
    const { property, type } = this.props;

    const qaDatasetForLi = this._isTesting()
      ? { "data-qa-property": property }
      : {};

    return dom.li(
      {
        key: `${property}:${type}`,
        ...qaDatasetForLi,
      },
      Object.entries(this.props).map(([key, value]) => {
        const qaDatasetForField = this._isTesting()
          ? {
              "data-qa-key": key,
              "data-qa-value": JSON.stringify(value),
            }
          : {};

        return dom.div(
          {
            key,
            ...qaDatasetForField,
          },
          `${key}:${JSON.stringify(value)}`
        );
      })
    );
  }
}

module.exports = IssueItem;
