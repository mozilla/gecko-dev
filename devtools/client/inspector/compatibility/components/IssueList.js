/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");

const Types = require("../types");

const IssueItem = createFactory(require("./IssueItem"));

class IssueList extends PureComponent {
  static get propTypes() {
    return {
      issues: PropTypes.arrayOf(PropTypes.shape(Types.issue)).isRequired,
    };
  }

  render() {
    const { issues } = this.props;

    return dom.ul({}, issues.map(issue => IssueItem({ ...issue })));
  }
}

module.exports = IssueList;
