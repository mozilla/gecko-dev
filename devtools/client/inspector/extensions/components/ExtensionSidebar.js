/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { createFactory, PureComponent } = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const { connect } = require("devtools/client/shared/vendor/react-redux");

const ExtensionPage = createFactory(require("./ExtensionPage"));
const ObjectTreeView = createFactory(require("./ObjectTreeView"));
const ObjectValueGripView = createFactory(require("./ObjectValueGripView"));
const Types = require("../types");

/**
 * The ExtensionSidebar is a React component with 2 supported viewMode:
 * - an ObjectTreeView UI, used to show the JS objects
 *   (used by the sidebar.setObject WebExtensions APIs)
 * - an ObjectValueGripView UI, used to show the objects value grips
 *   (used by sidebar.setExpression WebExtensions APIs)
 * - an ExtensionPage UI used to show an extension page
 *   (used by the sidebar.setPage WebExtensions APIs).
 *
 * TODO: implement the ExtensionPage viewMode.
 */
class ExtensionSidebar extends PureComponent {
  static get propTypes() {
    return {
      id: PropTypes.string.isRequired,
      extensionsSidebar: PropTypes.object.isRequired,
      onExtensionPageMount: PropTypes.func.isRequired,
      onExtensionPageUnmount: PropTypes.func.isRequired,
      // Helpers injected as props by extension-sidebar.js.
      serviceContainer: PropTypes.shape(Types.serviceContainer).isRequired,
    };
  }

  render() {
    const {
      id,
      extensionsSidebar,
      onExtensionPageMount,
      onExtensionPageUnmount,
      serviceContainer,
    } = this.props;

    const {
      iframeURL,
      object,
      objectValueGrip,
      rootTitle,
      viewMode = "empty-sidebar",
    } = extensionsSidebar[id] || {};

    let sidebarContentEl;

    switch (viewMode) {
      case "object-treeview":
        sidebarContentEl = ObjectTreeView({ object });
        break;
      case "object-value-grip-view":
        sidebarContentEl = ObjectValueGripView({
          objectValueGrip,
          rootTitle,
          serviceContainer,
        });
        break;
      case "extension-page":
        sidebarContentEl = ExtensionPage({
          iframeURL,
          onExtensionPageMount,
          onExtensionPageUnmount,
        });
        break;
      case "empty-sidebar":
        break;
      default:
        throw new Error(`Unknown ExtensionSidebar viewMode: "${viewMode}"`);
    }

    const className = "devtools-monospace extension-sidebar inspector-tabpanel";

    return dom.div({
      id,
      className,
      style: {
        height: "100%",
      },
    }, sidebarContentEl);
  }
}

module.exports = connect(state => state)(ExtensionSidebar);
