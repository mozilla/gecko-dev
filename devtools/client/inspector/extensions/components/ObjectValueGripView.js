/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { createFactory, PureComponent } = require("devtools/client/shared/vendor/react");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");

const Accordion = createFactory(require("devtools/client/inspector/layout/components/Accordion"));

// TODO: Upgrade to the current version reps https://bugzilla.mozilla.org/show_bug.cgi?id=1494680
const reps = require("devtools/client/shared/components/reps/reps-old");
const Types = require("../types");

const { REPS, MODE } = reps;
const { Grip } = REPS;

const ObjectInspector = createFactory(reps.ObjectInspector);

class ObjectValueGripView extends PureComponent {
  static get propTypes() {
    return {
      rootTitle: PropTypes.string,
      objectValueGrip: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.object,
      ]).isRequired,
      // Helpers injected as props by extension-sidebar.js.
      serviceContainer: PropTypes.shape(Types.serviceContainer).isRequired,
    };
  }

  render() {
    const {
      objectValueGrip,
      serviceContainer,
      rootTitle,
    } = this.props;

    const objectInspectorProps = {
      autoExpandDepth: 1,
      mode: MODE.SHORT,
      // TODO: we disable focus since it's not currently working well in ObjectInspector.
      // Let's remove the property below when problem are fixed in OI.
      disabledFocus: true,
      roots: [{
        path: objectValueGrip && objectValueGrip.actor || JSON.stringify(objectValueGrip),
        contents: {
          value: objectValueGrip,
        },
      }],
      createObjectClient: serviceContainer.createObjectClient,
      releaseActor: serviceContainer.releaseActor,
      // TODO: evaluate if there should also be a serviceContainer.openLink.
    };

    if (objectValueGrip && objectValueGrip.actor) {
      Object.assign(objectInspectorProps, {
        onDOMNodeMouseOver: serviceContainer.highlightDomElement,
        onDOMNodeMouseOut: serviceContainer.unHighlightDomElement,
        onInspectIconClick(object, e) {
          // Stop the event propagation so we don't trigger ObjectInspector
          // expand/collapse.
          e.stopPropagation();
          serviceContainer.openNodeInInspector(object);
        },
        defaultRep: Grip,
      });
    }

    if (rootTitle) {
      return Accordion({
        items: [
          {
            component: ObjectInspector,
            componentProps: objectInspectorProps,
            header: rootTitle,
            opened: true,
          },
        ],
      });
    }

    return ObjectInspector(objectInspectorProps);
  }
}

module.exports = ObjectValueGripView;
