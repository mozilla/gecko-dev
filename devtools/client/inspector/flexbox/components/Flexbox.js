/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const {
  createElement,
  createFactory,
  Fragment,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const dom = require("devtools/client/shared/vendor/react-dom-factories");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const { getStr } = require("devtools/client/inspector/layout/utils/l10n");

loader.lazyGetter(this, "FlexContainerProperties", function() {
  return createFactory(require("./FlexContainerProperties"));
});
loader.lazyGetter(this, "FlexItemList", function() {
  return createFactory(require("./FlexItemList"));
});
loader.lazyGetter(this, "FlexItemSizingOutline", function() {
  return createFactory(require("./FlexItemSizingOutline"));
});
loader.lazyGetter(this, "FlexItemSizingProperties", function() {
  return createFactory(require("./FlexItemSizingProperties"));
});
loader.lazyGetter(this, "Header", function() {
  return createFactory(require("./Header"));
});

const Types = require("../types");

class Flexbox extends PureComponent {
  static get propTypes() {
    return {
      flexbox: PropTypes.shape(Types.flexbox).isRequired,
      flexContainer: PropTypes.shape(Types.flexContainer).isRequired,
      getSwatchColorPickerTooltip: PropTypes.func.isRequired,
      onHideBoxModelHighlighter: PropTypes.func.isRequired,
      onSetFlexboxOverlayColor: PropTypes.func.isRequired,
      onShowBoxModelHighlighterForNode: PropTypes.func.isRequired,
      onToggleFlexboxHighlighter: PropTypes.func.isRequired,
      setSelectedNode: PropTypes.func.isRequired,
    };
  }

  renderFlexContainerProperties() {
    return FlexContainerProperties({
      properties: this.props.flexContainer.properties,
    });
  }

  renderFlexItemList() {
    const {
      onHideBoxModelHighlighter,
      onShowBoxModelHighlighterForNode,
      setSelectedNode,
    } = this.props;
    const { flexItems } = this.props.flexContainer;

    return FlexItemList({
      flexItems,
      onHideBoxModelHighlighter,
      onShowBoxModelHighlighterForNode,
      setSelectedNode,
    });
  }

  renderFlexItemSizing() {
    const {
      color,
    } = this.props.flexbox;
    const {
      flexItems,
      flexItemShown,
      properties,
    } = this.props.flexContainer;

    const flexItem = flexItems.find(item => item.nodeFront.actorID === flexItemShown);
    if (!flexItem) {
      return null;
    }

    return createElement(Fragment, null,
      FlexItemSizingOutline({
        color,
        flexDirection: properties["flex-direction"],
        flexItem,
      }),
      FlexItemSizingProperties({
        flexDirection: properties["flex-direction"],
        flexItem,
      })
    );
  }

  render() {
    const {
      flexContainer,
      getSwatchColorPickerTooltip,
      onHideBoxModelHighlighter,
      onSetFlexboxOverlayColor,
      onShowBoxModelHighlighterForNode,
      onToggleFlexboxHighlighter,
      setSelectedNode,
    } = this.props;

    if (!flexContainer.actorID) {
      return (
        dom.div({ className: "devtools-sidepanel-no-result" },
          getStr("flexbox.noFlexboxeOnThisPage")
        )
      );
    }

    const {
      flexItems,
      flexItemShown,
    } = flexContainer;

    return (
      dom.div({ id: "layout-flexbox-container" },
        Header({
          flexContainer,
          getSwatchColorPickerTooltip,
          onHideBoxModelHighlighter,
          onSetFlexboxOverlayColor,
          onShowBoxModelHighlighterForNode,
          onToggleFlexboxHighlighter,
          setSelectedNode,
        }),
        !flexItemShown && flexItems.length > 0 ? this.renderFlexItemList() : null,
        flexItemShown ? this.renderFlexItemSizing() : null,
        !flexItemShown ? this.renderFlexContainerProperties() : null
      )
    );
  }
}

module.exports = Flexbox;
