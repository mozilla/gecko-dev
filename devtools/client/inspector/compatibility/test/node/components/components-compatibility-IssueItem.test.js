/* Any copyright is dedicated to the Public Domain.
   http://creativecommons.org/publicdomain/zero/1.0/ */

"use strict";

/**
 * Unit tests for the IssueItem component.
 */

const { shallow } = require("enzyme");
const React = require("react");

const MDNCompatibility = require("devtools/client/inspector/compatibility/lib/MDNCompatibility");
const IssueItem = React.createFactory(
  require("devtools/client/inspector/compatibility/components/IssueItem")
);

describe("IssueItem component", () => {
  it("renders an issue of CSS property", () => {
    const item = shallow(
      IssueItem({
        type: MDNCompatibility.ISSUE_TYPE.CSS_PROPERTY,
        property: "test-property",
        url: "test-url",
        deprecated: false,
        experimental: false,
        unsupportedBrowsers: [
          { id: "firefox", name: "Firefox", version: "70", status: "nightly" },
        ],
      })
    );
    expect(item).toMatchSnapshot();
  });
});
