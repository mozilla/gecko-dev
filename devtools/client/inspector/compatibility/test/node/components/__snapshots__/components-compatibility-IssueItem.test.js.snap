// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueItem component renders an issue of CSS property 1`] = `
<li
  data-qa-property="test-property"
  key="test-property:CSS_PROPERTY"
>
  <div
    data-qa-key="type"
    data-qa-value="\\"CSS_PROPERTY\\""
    key="type"
  >
    type:"CSS_PROPERTY"
  </div>
  <div
    data-qa-key="property"
    data-qa-value="\\"test-property\\""
    key="property"
  >
    property:"test-property"
  </div>
  <div
    data-qa-key="url"
    data-qa-value="\\"test-url\\""
    key="url"
  >
    url:"test-url"
  </div>
  <div
    data-qa-key="deprecated"
    data-qa-value="false"
    key="deprecated"
  >
    deprecated:false
  </div>
  <div
    data-qa-key="experimental"
    data-qa-value="false"
    key="experimental"
  >
    experimental:false
  </div>
  <div
    data-qa-key="unsupportedBrowsers"
    data-qa-value="[{\\"id\\":\\"firefox\\",\\"name\\":\\"Firefox\\",\\"version\\":\\"70\\",\\"status\\":\\"nightly\\"}]"
    key="unsupportedBrowsers"
  >
    unsupportedBrowsers:[{"id":"firefox","name":"Firefox","version":"70","status":"nightly"}]
  </div>
</li>
`;
